pkgs/build-support/docker/default.nix

```diff
@@ -866,4 +869,96 @@ rec {
+  buildNixShellImage =
+    # Use the environment of this derivation
+    { name ? drv.name + "-env"
+    , drv
+    , tag ? null
+    # Files to put on the image (a nix store path or list of paths).
+    , contents ? []
+    # uid/gid to run as inside the container
+    # Some compilers don't like to run as root
+    , uid ? 1000, gid ? 1000
+  }:
+    assert lib.assertMsg (! (drv.__structuredAttrs or false))
+      "buildNixShellImage does not work with the derivation ${drv.drvPath} because it uses __structuredAttrs";
+    let
+
+      # A mapping from output name to the nix store path where they should end up
+      outputPaths = lib.genAttrs drv.outputs (output: builtins.unsafeDiscardStringContext drv.${output}.outPath);
+
+      # Additional binaries available in this image, both as the initial
+      # command of the arguments, and within a bash
+      binaryPath = lib.makeBinPath [ bashInteractive builder coreutils ];
+
+      setupPackage = runCommandNoCC "setup" {
+        # Mimics a standard nix-shell
+        rc = ''
+          source $stdenv/setup
+          set +e
+          runHook shellHook
+          export PATH=$PATH:${binaryPath}
+        '';
+        passAsFile = [ "rc" ];
+      } ''
+        mkdir -p $out/{etc/ssl/certs,tmp,bin,usr/bin,home/user/{outputs,build}}
+        ln -s ${bashInteractive}/bin/sh $out/bin/sh
+        ln -s ${coreutils}/bin/env $out/usr/bin/env
+        cp "$rcPath" $out/home/user/.bashrc
+        ln -s "${cacert}/etc/ssl/certs/ca-bundle.crt" $out/etc/ssl/certs/ca-certificates.crt
+
+        ${lib.concatStringsSep "\n" (lib.mapAttrsToList (name: path: ''
+          ln -s ${path} $out/home/user/outputs/${name}
+        '') outputPaths)}
+      '';
+
+      # Support for passAsFile
+      passAsFiles = lib.listToAttrs (map (attr:
+        lib.nameValuePair "${attr}Path" (writeText "pass-as-text-${attr}" drv.drvAttrs.${attr})
+      ) (drv.drvAttrs.passAsFile or []));
+
+      # We can't just use `toString` on all derivation attributes because that
+      # would not put path literals in the closure. So we explicitly copy
+      # those into the store here
+      stringValue = value: if builtins.typeOf value == "path" then "${value}"
+        else if builtins.isList value then toString (map stringValue value)
+        else toString value;
+
+      # Environment variables set in the image
+      envVars = lib.mapAttrs (name: stringValue) drv.drvAttrs // outputPaths // passAsFiles // {
+        # These are needed for stdenv setup to work
+        NIX_STORE = storeDir;
+        NIX_BUILD_TOP = "/home/user/build";
+
+        HOME = "/home/user";
+        TMPDIR = "/tmp";
+        PATH = binaryPath;
+      };
+
+      # A binary that calls the command to build the derivation
+      builder = writeShellScriptBin "buildDerivation" ''
+        exec ${lib.escapeShellArg (stringValue drv.drvAttrs.builder)} ${lib.escapeShellArgs (map stringValue drv.drvAttrs.args)}
+      '';
+
+    in buildLayeredImage {
```

Comment:
I want to override some settings in this config. I also want to use `streamLayeredImage`. So really it makes sense to just return the attrset directly and let the user decide where the image is actually realized.

Reply:
You could use `(buildLayeredImage args).stream` to solve that part of the problem.
I feel the increasing number of interfaces for building docker images is starting to become a problem, so I'm not excited about having a function that returns arguments for another function. I'd like to experiment with a more powerful interface during Ocean Sprint, so I'll know more in a month.